code/lib/cli/src/automigrate/fixes/prompt-remove-react.ts

```diff
@@ -0,0 +1,133 @@
+            })
+          ).map((f) => relative(absoluteDirectory, f)),
+          directory,
+          titlePrefix,
+        };
+      })
+    )
+  ).reduce<boolean>((acc, { files }, i) => {
+    const filteredEntries = files.filter((s) => !s.endsWith('.mdx'));
+    if (filteredEntries.length < files.length) {
+      return true;
+    }
+    return acc;
+  }, false);
+  return result;
+}
+
+export const removeReactDependency: Fix<Options> = {
+  id: 'remove-react-dependency',
+
+  async check({ packageManager, mainConfig, storybookVersion, configDir }) {
+    let hasMDX = false;
+    let hasEssentials = false;
+    let hasDocs = false;
+
+    // when the user is using the react renderer, we should not prompt them to remove react
+    const frameworkPackageName = getFrameworkPackageName(mainConfig);
+    if (frameworkPackageName?.includes('react') || frameworkPackageName?.includes('nextjs')) {
+      return null;
+    }
+
+    // if the user has no dependency on react, we can skip this fix
+    const packageJson = await packageManager.retrievePackageJson();
+    if (
+      !packageJson?.dependencies?.['react'] &&
+      !packageJson?.peerDependencies?.['react'] &&
+      !packageJson?.devDependencies?.['react']
+    ) {
+      return null;
+    }
+
+    // do not prompt to remove react for older versions of storybook
+    if (!semver.gte(storybookVersion, '8.0.0')) {
+      return null;
+    }
+
+    const { addons } = mainConfig;
+    hasEssentials = !!(
+      addons &&
+      addons.find((addon) =>
+        typeof addon === 'string'
+          ? addon.endsWith('@storybook/addon-essentials')
+          : addon.name.endsWith('@storybook/addon-essentials')
+      )
+    );
+    hasDocs = !!(
+      addons &&
+      addons.find((addon) =>
+        typeof addon === 'string'
+          ? addon.endsWith('@storybook/addon-docs')
+          : addon.name.endsWith('@storybook/addon-docs')
+      )
+    );
+
+    hasMDX = !!(await detectMDXEntries(mainConfig.stories, configDir || process.cwd()));
+
+    return {
+      hasMDX,
+      hasEssentials,
+      hasDocs,
+    };
+  },
+  prompt({ hasMDX, hasDocs, hasEssentials }) {
+    const addons = [hasEssentials ? 'essentials' : '', hasDocs ? 'docs' : ''].filter(Boolean);
+    const addonReasonText =
+      addons.length > 0 ? `, because you are using ${addons.join(' and ')}` : '';
+
+    const start = dedent`
+      We detected that your project has a dependency for "react" that it might not need.
+      Nothing breaks by having it, you can safely ignore this message, if you wish.
+
+      Storybook asked you to add "react" as a direct dependency in the past${addonReasonText}.
+      However, since version 8.0, Storybook no longer requires you to provide "react" as a dependency.
+    `;
+
+    if (hasMDX) {
+      const mdxSuggestion = dedent`
+        As you are using '.mdx'-files, it might be reasonable to keep the dependency.
+      `;
+      return [start, mdxSuggestion].join('\n\n');
+    }
```

Comment:
The `detectMDXEntries` is pretty heavy and requires some heavy-lifting file system lookups. I don't think it's worth it just to log this one-liner to the user. What is the worst that can happen? Things might break and the user has to read `react` and `react-dom`. I think that's acceptable.

Reply:
When the user uses `.mdx` files, they will very likely want to keep the react dependency.

I pulled this code from the override preset, it costs like <80ms. Is that heavy lifting file system lookups to you?

Reply:
I don't know in which particular project you did the measurement, but I guess that the automigration might be blocking in bigger projects.

Also, I don't quite understand what you want to achieve:

- You only scan for mdx files, if they are part of the `stories` glob. If they are not defined in the stories glob, you are not checking for them. So why not directly just search for `mdx` in the stories glob pattern? I don't think file system lookups are necessary.
- What about MDX usage outside of Storybook. This case isn't covered here, since if the `stories` glob pattern doesn't mention `mdx` files, they also subsequently will not be found by glob. So if you want to cover this scenario, you don't have to consider the `stories` glob at all but create a simple glob, which considers mdx files and pass it directly to the `glob` function.

Reply:
I also don't think this check is necessary at all. Early results in https://github.com/storybookjs/storybook/pull/24881 shows that it's perfectly valid to have MDX files - even ones importing React components - in a project without `react` and `react-dom`. It just works. TypeScript in the IDE might complain - I haven't tested that out yet - but I don't think that warrants this additional message.